tests/expose_test.go

```diff
@@ -407,6 +407,77 @@ var _ = Describe("Expose", func() {
 
 				Eventually(getStatus, 60, 1).Should(Equal(k8sv1.PodSucceeded))
 			})
+
+			It("Should verify the exposed service is functional before and after VM restart.", func() {
+				vm_obj := vm
+
+				By("Getting back the service's allocated cluster IP.")
+				svc, err := virtClient.CoreV1().Services(vm_obj.Namespace).Get(serviceName, k8smetav1.GetOptions{})
+				Expect(err).ToNot(HaveOccurred())
+				serviceIP := svc.Spec.ClusterIP
+
+				By("Starting a pod which tries to reach the VMI via ClusterIP.")
+				job := tests.NewHelloWorldJob(serviceIP, servicePort)
+				job, err = virtClient.CoreV1().Pods(vm_obj.Namespace).Create(job)
+				Expect(err).ToNot(HaveOccurred())
+
+				By("Waiting for the pod to report a successful connection attempt.")
+				getStatus := func() k8sv1.PodPhase {
+					pod, err := virtClient.CoreV1().Pods(job.Namespace).Get(job.Name, k8smetav1.GetOptions{})
+					Expect(err).ToNot(HaveOccurred())
+					return pod.Status.Phase
+				}
+
+				Eventually(getStatus, 60, 1).Should(Equal(k8sv1.PodSucceeded))
+
+				By("Stopping the running VM.")
+				virtctl := tests.NewRepeatableVirtctlCommand("stop", "--namespace", vm.Namespace, vm.Name)
+				err = virtctl()
+				Expect(err).ToNot(HaveOccurred())
+
+				By("Verifying the VMI is not running.")
+				var vmi *v1.VirtualMachineInstance
+				Eventually(func() bool {
+					vmi, err = virtClient.VirtualMachineInstance(vm_obj.Namespace).Get(vm_obj.Name, &k8smetav1.GetOptions{})
+					Expect(err).ToNot(HaveOccurred())
+					return vmi.Status.Phase != v1.Running
+				}, 120*time.Second, 1*time.Second).Should(BeTrue())
+
+				By("Starting the VM.")
+				virtctl = tests.NewRepeatableVirtctlCommand("start", "--namespace", vm.Namespace, vm.Name)
+				err = virtctl()
+				Expect(err).ToNot(HaveOccurred())
+
+				By("Verifying the VMI is running.")
+				Eventually(func() bool {
+					vmi, err = virtClient.VirtualMachineInstance(vm_obj.Namespace).Get(vm_obj.Name, &k8smetav1.GetOptions{})
+					Expect(err).ToNot(HaveOccurred())
+					return vmi.Status.Phase == v1.Running
+				}, 120*time.Second, 1*time.Second).Should(BeTrue())
+
+				By("Creating a TCP server on the VM.")
+				generateHelloWorldServer(vmi, virtClient, testPort, "tcp")
+
+				By("Repeating the sequence as prior to restarting the VM: Connect to exposed ClusterIP service.")
+				By("Getting back the service's allocated cluster IP.")
+				svc, err = virtClient.CoreV1().Services(vm_obj.Namespace).Get(serviceName, k8smetav1.GetOptions{})
```

Comment:
Agreed.
I mistakenly assumed the exposed service is also restarted, along with the VMI, and therefore should also be retrieved again. That doesn't make sense, however, because it's actually a cluster service.